src/mlpack/core/data/load_impl.hpp

```diff
@@ -120,11 +120,8 @@ bool Load(const std::string& filename,
 
   if (extension == "csv" || extension == "tsv")
   {
-#if (ARMA_VERSION_MAJOR == 9 && ARMA_VERSION_MINOR >= 800)
-    loadType = arma::diskio::guess_file_type_internal(stream);
-#else
     loadType = arma::diskio::guess_file_type(stream);
-#endif
+
```

Comment:
@conradsnicta so, while this will avoid the use of a specifically marked `internal` function, let's talk about the actual needs here.  This particular function, `Load()`, attempts to detect the file type based on the name of the file.  So, e.g., `filename.csv` and `filename.txt` can have their type auto-detected---but, as I'm sure you know from dealing with real-world data, `.csv` indicates only that the file *might* be CSV.  It could also be TSV or even just space-separated values.  So, in the past, we used `guess_file_type()` here to ensure that we would load files exactly like Armadillo does.  Another important reason for detecting the file type based on extension is that (at least a long time ago) Armadillo's auto-detection would default to `raw_binary` if `guess_file_type()` didn't find a header for other types.  So, using the extension is actually important here, because loading any strange file as `arma_binary` gives really weird and unexpected results in the mlpack context.

So, from this perspective, it's important to reuse Armadillo's functionality exactly for three reasons:

1. So we don't get bug reports if mlpack behaves slightly differently than Armadillo (i.e. "Armadillo loads this as TSV but mlpack claims it's invalid").

2. So that if we detect something as a valid CSV, that we can expect that Armadillo will succeed in loading it.

3. So we don't have to maintain all kinds of additional code that does what Armadillo already does.

With this in mind, what is your thought here for how to handle this?  Changing the external API for `Load()` is something I'd really rather avoid, since that has been stable for a decade.

Reply:
The Load() function in mlpack is not the same as the Mat::load() function in Armadillo, so I'm not sure there should be an expectation that they would (or even should) work exactly the same and handle the same file types.  mlpack could very well handle more file types, or have better detection.

I understand the pros of reusing functionality already present elsewhere, however the diskio::guess_file_type() function was never designed (and never documented) to be used outside of Armadillo.  I don't want to accidentally break mlpack because an internal Armadillo function was refactored, extended, or otherwise changed. That's the entire point of separating internal and external API, where the latter carries a guarantee of stability.

However, your point about staying compatible with Armadillo file loading/saving is fair.  To address that point, I still don't recommend using diskio::guess_file_type().  Instead, I recommend having dedicated code in mlpack for detecting each possible file type that mlpack wants to handle, using possibly better/more robust heuristics than Armadillo's internal detection heuristics, or even going beyond the file types that Armadillo handles.  Then for each detected type, explicitly specifying that file type when calling Mat::load(). As an example, if CSV is detected, then the "csv_ascii" file type is given to Mat::load().  Armadillo will then assume that the file is indeed CSV and load it accordingly, completely bypassing diskio::guess_file_type().

Related note: in Armadillo 9.900, there's optional handling of CSV files with headers, which is not handled by diskio::guess_file_type() at all.  It might be useful for mlpack to make use of this new functionality.  See the "csv_name" form of [.save()/.load()](http://arma.sourceforge.net/docs.html#save_load_mat).

There's already a related pull request in mlpack, with (almost finished) work on an mlpack specific approach for auto detection of file types: https://github.com/mlpack/mlpack/pull/2135.  I recommend getting this over the line.

Reply:
Given what you wrote, I have to agree, and #2135 does seem like the right way to go here.  I'll resurrect that as another branch and we can get it merged soon.

I would like to find a way to handle the CSV headers soon too---that is nice support in Armadillo!